Add `storage.wal.fsync.latency` to metrics (#19425)
cockroachdb/docs: src/current/_includes/v25.1/essential-metrics.md

| <a id="capacity-available"></a>capacity.available                                  | capacity.available                                           | Available storage capacity                                   | This metric gives available storage capacity. Measurements should comply with the following rule: CockroachDB storage volumes should not be utilized more than 60% (40% free space). |
| capacity.used                                       | capacity.used                                                | Used storage capacity                                        | This metric gives used storage capacity. Measurements should comply with the following rule: CockroachDB storage volumes should not be utilized more than 60% (40% free space). |

> | <a id="storage-wal-fsync-latency"></a>storage.wal.fsync.latency | {% if include.deployment == 'self-hosted' %}storage.wal.fsync.latency |{% elsif include.deployment == 'advanced' %}storage.wal.fsync.latency |{% endif %} This metric reports the latency of writes to the [WAL]({% link {{ page.version.version }}/architecture/storage-layer.md %}#memtable-and-write-ahead-log). | If this value is greater than `100ms`, it is an indication of a [disk stall]({% link {{ page.version.version }}/cluster-setup-troubleshooting.md %}#disk-stalls). To mitigate the effects of disk stalls, consider deploying your cluster with [WAL failover]({% link {{ page.version.version }}/wal-failover.md %}) configured. |

Reviewer: Is the `if` condition necessary? It looks like the content is the same, but maybe I'm missing something.

Author: i'm not sure. i don't usually work on stuff in this file so this is cargo-culted from an existing row in this table to keep to the same format, on the principle that if someone yells because it breaks i'll come back and fix (this PR is already long in tooth, my fault for starting it before PTO)

- meta: i don't think liquid logic in this table is the best way to express "this metric is available in this delivery method" is the best way to do this but i'll litigate that separately from this PR. would love to have some more principled way to express this (ideally the product would have more consistency across delivery mechanisms to obviate the need to express so much logic of this type)

| ----------------------------------------------------- | {% if include.deployment == 'self-hosted' %}------ |{% elsif include.deployment == 'advanced' %}---- |{% endif %} ------------------------------------------------------------ | ------------------------------------------------------------ |
| leases.transfers.success                            | leases.transfers.success                                     | Number of successful lease transfers                         | A high number of [lease](architecture/replication-layer.html#leases) transfers is not a negative or positive signal, rather it is a reflection of the elastic cluster activities. For example, this metric is high during cluster topology changes. A high value is often the reason for NotLeaseHolderErrors which are normal and expected during rebalancing. Observing this metric may provide a confirmation of the cause of such errors. |

> | rebalancing_lease_transfers | rebalancing.lease.transfers | Counter of the number of [lease transfers]({% link {{ page.version.version }}/architecture/replication-layer.md %}#epoch-based-leases-table-data) that occur during replica rebalancing. | [XXX](): USAGE??? |

Author: hi @kvoli 

 i've been asked to add some docs for the following metrics which appear to be replication-related (via https://cockroachlabs.atlassian.net/browse/DOC-11996):

- `rebalancing_lease_transfers`
- `rebalancing_range_rebalances`
- `rebalancing_replicas_queriespersecond`

i've taken a shot at writing descriptions for each which i'd really appreciate your feedback on

there is also a 'usage' column which appears to explain "what is this metric for / why should i watch it?", in the case of `rebalancing_replicas_queriespersecond` i tried to write this based on an old slack convo of yours i found in glean - please let me know what you think, happy to update

for the other two metrics, I wasn't sure what to write w.r.t. usage, can you please explain what these metrics are used for / why a user should pay attention to them? my sense is something like: too much range rebalancing/lease transfer =  bad, but i'm guessing there is more to it than that!

Reviewer 2: Hey @rmloveland,

For `rebalancing_lease_transfers`, these are lease transfers specific to a certain component which looks for store level load imbalance when looking at either QPS (`rebalancing.queriespersecond`) or CPU (`rebalancing.cpunanospersecond`) depending on what the `kv.allocator.load_based_rebalancing.objective` is set to (`qps` or `cpu`). Likewise for `rebalancing_range_rebalances`, except the action is moving the replicas of a range (potentially including the leaseholder), instead of just the lease.

For `rebalancing_replicas_queriespersecond` this is a histogram (also see `rebalancing_replicas_cpunanospersecond`) which instead of only maintaining a gauge of the stores currently reported QPS or CPU, it maintains buckets so a user could query the PXX replica's  QPS, or CPU. If we don't already have docs for the non-histogram variants`rebalancing.queriespersecond`, `rebalancing.cpunanospersecond`, I'd suggest including these instead.

I generally use and have seen the two rebalance action counter metrics `rebalancing_(lease_transfers|range_rebalances)` used to identify when there has been more rebalancing activity triggered by imbalance (QPS or CPU) between stores.

For the rebalancing load metrics, since these map directly to what the rebalancing algorithm is looking at when attempting to balance load, I have seen these used to identify the efficiency rebalancing and check for store level imbalances (in cases where there is a hardware level load imbalance).

> my sense is something like: too much range rebalancing/lease transfer = bad, but i'm guessing there is more to it than that!

Partially answered above, if these are high (when the count is rated), it can indicate thrashing. Where stores continuously move replicas and leases among eachother without actually improving the overall load distribution in the medium-long term. If there is no activity and the stores remain imbalanced, this is an indicator that no rebalance activity is being undertaken to fix the imbalance, potentially signalling an issue to the user.

Author: > Hey @rmloveland,
> 
> For `rebalancing_lease_transfers`, these are lease transfers specific to a certain component which looks for store level load imbalance when looking at either QPS (`rebalancing.queriespersecond`) or CPU (`rebalancing.cpunanospersecond`) depending on what the `kv.allocator.load_based_rebalancing.objective` is set to (`qps` or `cpu`). Likewise for `rebalancing_range_rebalances`, except the action is moving the replicas of a range (potentially including the leaseholder), instead of just the lease.
> 
> For `rebalancing_replicas_queriespersecond` this is a histogram (also see `rebalancing_replicas_cpunanospersecond`) which instead of only maintaining a gauge of the stores currently reported QPS or CPU, it maintains buckets so a user could query the PXX replica's QPS, or CPU. If we don't already have docs for the non-histogram variants`rebalancing.queriespersecond`, `rebalancing.cpunanospersecond`, I'd suggest including these instead.
>
> I generally use and have seen the two rebalance action counter metrics `rebalancing_(lease_transfers|range_rebalances)` used to identify when there has been more rebalancing activity triggered by imbalance (QPS or CPU) between stores.
> 
> For the rebalancing load metrics, since these map directly to what the rebalancing algorithm is looking at when attempting to balance load, I have seen these used to identify the efficiency rebalancing and check for store level imbalances (in cases where there is a hardware level load imbalance).
> 
> > my sense is something like: too much range rebalancing/lease transfer = bad, but i'm guessing there is more to it than that!
> 
> Partially answered above, if these are high (when the count is rated), it can indicate thrashing. Where stores continuously move replicas and leases among eachother without actually improving the overall load distribution in the medium-long term. If there is no activity and the stores remain imbalanced, this is an indicator that no rebalance activity is being undertaken to fix the imbalance, potentially signalling an issue to the user.

Thanks, in the latest commit I have tried to:

- update `rebalancing_lease_transfers` and `rebalancing_range_rebalances` to incorporate info from your response above in both the 'Description' and 'Usage' columns, as well as adding a "see also: the non-histogram variant X" to the metrics that had such variants
- make sure we had docs for the non-histogram variants you mentioned `rebalancing.queriespersecond` (already existed) and `rebalancing.cpunanospersecond` (wrote new)

PTAL and let me know what you think

PS i'm updating this to a "real PR" now since I think these will be the bulk of the changes (the WAL failover metric i'll get a separate review from an engineer on Storage  team once we're happy with these replication-related metrics)

| ----------------------------------------------------- | {% if include.deployment == 'self-hosted' %}------ |{% elsif include.deployment == 'advanced' %}---- |{% endif %} ------------------------------------------------------------ | ------------------------------------------------------------ |
| leases.transfers.success                            | leases.transfers.success                                     | Number of successful lease transfers                         | A high number of [lease](architecture/replication-layer.html#leases) transfers is not a negative or positive signal, rather it is a reflection of the elastic cluster activities. For example, this metric is high during cluster topology changes. A high value is often the reason for NotLeaseHolderErrors which are normal and expected during rebalancing. Observing this metric may provide a confirmation of the cause of such errors. |

> | rebalancing_lease_transfers | rebalancing.lease.transfers | Counter of the number of [lease transfers]({% link {{ page.version.version }}/architecture/replication-layer.md %}#epoch-based-leases-table-data) that occur during replica rebalancing. These lease transfers are tracked by a component which looks for [store-level]({% link {{ page.version.version }}/cockroach-start.md %}#store) load imbalance when looking at either QPS (`rebalancing.queriespersecond`) or CPU (`rebalancing.cpunanospersecond`) depending on the value of the `kv.allocator.load_based_rebalancing.objective` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}#setting-kv-allocator-load-based-rebalancing-objective). | Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it indicates more rebalancing activity is taking place due to load imbalance between stores. |

Reviewer: ```suggestion
| rebalancing_lease_transfers | rebalancing.lease.transfers | Counter of the number of [lease transfers]({% link {{ page.version.version }}/architecture/replication-layer.md %}#epoch-based-leases-table-data) that occur during replica rebalancing. These lease transfers are tracked by a component that looks for a [store-level]({% link {{ page.version.version }}/cockroach-start.md %}#store) load imbalance of either QPS (`rebalancing.queriespersecond`) or CPU usage (`rebalancing.cpunanospersecond`), depending on the value of the `kv.allocator.load_based_rebalancing.objective` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}#setting-kv-allocator-load-based-rebalancing-objective). | Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it indicates that more rebalancing activity is taking place due to load imbalance between stores. |
```

> | rebalancing_lease_transfers | rebalancing.lease.transfers | Counter of the number of [lease transfers]({% link {{ page.version.version }}/architecture/replication-layer.md %}#epoch-based-leases-table-data) that occur during replica rebalancing. These lease transfers are tracked by a component which looks for [store-level]({% link {{ page.version.version }}/cockroach-start.md %}#store) load imbalance when looking at either QPS (`rebalancing.queriespersecond`) or CPU (`rebalancing.cpunanospersecond`) depending on the value of the `kv.allocator.load_based_rebalancing.objective` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}#setting-kv-allocator-load-based-rebalancing-objective). | Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it can indicate thrashing where stores continuously move replicas and leases among each other without actually improving the overall load distribution in the medium to long term. If there is no activity on this metric and the stores remain imbalanced, this is an indicator that no rebalance activity is being undertaken to fix the imbalance, potentially signalling an issue. |

Reviewer 2: Consider shortening the latter parts regarding thrashing and no action being taken during imbalance into something like:

```
... it indicates more rebalancing activity is taking place due to load imbalance between stores.
```

The additional thrashing and no activity parts are useful in a support context (and in sharing context in our earlier thread).

wdyt?

Author: sgtm! I can see the stuff about thrashing, etc., not necessarily being helpful in docs since ... at that point you may need help from support, etc.

shortened based on your feedback to:

> Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it indicates more rebalancing activity is taking place due to load imbalance between stores.

> | rebalancing_range_rebalances | {% if include.deployment == 'self-hosted' %}rebalancing.range.rebalances | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter of the number of [load-based range rebalances]({% link {{ page.version.version }}/architecture/replication-layer.md %}#load-based-replica-rebalancing). This range movement is tracked by a component which looks for [store-level]({% link {{ page.version.version }}/cockroach-start.md %}#store) load imbalance when looking at either QPS (`rebalancing.queriespersecond`) or CPU (`rebalancing.cpunanospersecond`) depending on the value of the `kv.allocator.load_based_rebalancing.objective` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}#setting-kv-allocator-load-based-rebalancing-objective). | Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it indicates more rebalancing activity is taking place due to load imbalance between stores. |

Reviewer: ```suggestion
| rebalancing_range_rebalances | {% if include.deployment == 'self-hosted' %}rebalancing.range.rebalances | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter of the number of [load-based range rebalances]({% link {{ page.version.version }}/architecture/replication-layer.md %}#load-based-replica-rebalancing). This range movement is tracked by a component that looks for [store-level]({% link {{ page.version.version }}/cockroach-start.md %}#store) load imbalance of either QPS (`rebalancing.queriespersecond`) or CPU usage (`rebalancing.cpunanospersecond`), depending on the value of the `kv.allocator.load_based_rebalancing.objective` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}#setting-kv-allocator-load-based-rebalancing-objective). | Used to identify when there has been more rebalancing activity triggered by imbalance between stores (of QPS or CPU). If this is high (when the count is rated), it indicates that more rebalancing activity is taking place due to load imbalance between stores. |
```

> | rebalancing_replicas_queriespersecond | {% if include.deployment == 'self-hosted' %}rebalancing.replicas.queriespersecond | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter for the kv-level requests received per second by a given [store]({% link {{ page.version.version }}/cockroach-start.md %}#store). The store aggregates all of the CPU and QPS stats across all its replicas and then creates a histogram which maintains buckets so you can query, e.g., the P95 replica's QPS or CPU. | A high value of this metric could indicate that one of the store's replicas is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also: `rebalancing_replicas_cpunanospersecond`. |

Reviewer: ```suggestion
| rebalancing_replicas_queriespersecond | {% if include.deployment == 'self-hosted' %}rebalancing.replicas.queriespersecond | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter of the KV-level requests received per second by a given [store]({% link {{ page.version.version }}/cockroach-start.md %}#store). The store aggregates all of the CPU and QPS stats across all its replicas and then creates a histogram that maintains buckets that can be queried for, e.g., the P95 replica's QPS or CPU. | A high value of this metric could indicate that one of the store's replicas is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also: `rebalancing_replicas_cpunanospersecond`. |
```

> | rebalancing_replicas_queriespersecond | {% if include.deployment == 'self-hosted' %}rebalancing.replicas.queriespersecond | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter for the kv-level requests received per second by a given [replica]({% link {{ page.version.version }}/architecture/overview.md %}#architecture-replica). This metric is a histogram which maintains buckets so you can query, e.g., the P95 replica's QPS or CPU. | A high value of this metric for a particular replica could indicate that the replica is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also: `rebalancing_replicas_cpunanospersecond`. |

Reviewer 2: For the histogram metrics, the metric is per-store not per-replica. The store aggregates all of its replica's CPU and QPS stats and then creates a histogram.  Consider removing the "particular replica" wording.

Author: Ah ok thanks for clarifying that. I have updated the description to add the info re: "the store aggregates ..." from your comment as well as remove the "particular replica" wording.

Updated to the following, PTAL and suggest edits as needed!

> Counter for the {kv-level requests received,CPU nanoseconds of execution time} per second by a given [store]({% link {{ page.version.version }}/cockroach-start.md %}#store). The store aggregates all of the CPU and QPS stats across all its replicas and then creates a histogram which maintains buckets so you can query, e.g., the P95 replica's QPS or CPU. | A high value of this metric could indicate that one of the store's replicas is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also: `rebalancing_replicas_cpunanospersecond`. |

> | rebalancing_replicas_cpunanospersecond | {% if include.deployment == 'self-hosted' %}rebalancing.replicas.cpunanospersecond | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter for the CPU nanoseconds of execution time per second by a given [store]({% link {{ page.version.version }}/cockroach-start.md %}#store). The store aggregates all of the CPU and QPS stats across all its replicas and then creates a histogram which maintains buckets so you can query, e.g., the P95 replica's QPS or CPU. | A high value of this metric could indicate that one of the store's replicas is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also the non-histogram variant: `rebalancing.cpunanospersecond`. |

Reviewer: ```suggestion
| rebalancing_replicas_cpunanospersecond | {% if include.deployment == 'self-hosted' %}rebalancing.replicas.cpunanospersecond | {% elsif include.deployment == 'advanced' %}NOT AVAILABLE |{% endif %} Counter of the CPU nanoseconds of execution time per second by a given [store]({% link {{ page.version.version }}/cockroach-start.md %}#store). The store aggregates all of the CPU and QPS stats across all its replicas and then creates a histogram that maintains buckets that can be queried for, e.g., the P95 replica's QPS or CPU. | A high value of this metric could indicate that one of the store's replicas is part of a [hot range]({% link {{ page.version.version }}/understand-hotspots.md %}#hot-range). See also the non-histogram variant: `rebalancing.cpunanospersecond`. |
```
